pkg/controller/nstemplateset/nstemplateset_controller.go

```diff
@@ -0,0 +1,260 @@
+	errs "github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var log = logf.Log.WithName("controller_nstemplateset")
+
+const (
+	// Status condition reasons
+	unableToProvisionReason          = "UnableToProvision"
+	unableToProvisionNamespaceReason = "UnableToProvisionNamespace"
+	provisioningNamespaceReason      = "ProvisioningNamespace"
+	provisioningReason               = "Provisioning"
+	provisionedReason                = "Provisioned"
+)
+
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileNSTemplateSet{
+		client:        mgr.GetClient(),
+		scheme:        mgr.GetScheme(),
+		applyTemplate: applyTemplate,
+	}
+}
+
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("nstemplateset-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource
+	err = c.Watch(&source.Kind{Type: &toolchainv1alpha1.NSTemplateSet{}}, &handler.EnqueueRequestForObject{}, predicate.GenerationChangedPredicate{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to secondary resource
+	enqueueRequestForOwner := &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &toolchainv1alpha1.NSTemplateSet{},
+	}
+	if err := c.Watch(&source.Kind{Type: &corev1.Namespace{}}, enqueueRequestForOwner); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileNSTemplateSet{}
+
+type ReconcileNSTemplateSet struct {
+	client        client.Client
+	scheme        *runtime.Scheme
+	applyTemplate func(client.Client, toolchainv1alpha1.Namespace, map[string]string) error
+}
+
+// Reconcile reads that state of the cluster for a NSTemplateSet object and makes changes based on the state read
+// and what is in the NSTemplateSet.Spec
+func (r *ReconcileNSTemplateSet) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling NSTemplateSet")
+
+	// Fetch the NSTemplateSet instance
+	nsTmplSet := &toolchainv1alpha1.NSTemplateSet{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, nsTmplSet)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	result, err := r.ensureNamespaces(reqLogger, nsTmplSet)
+	if err != nil || result.Requeue == true {
+		return result, err
+	}
+	return result, r.setStatusReady(nsTmplSet)
+}
+
+func (r *ReconcileNSTemplateSet) ensureNamespaces(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet) (reconcile.Result, error) {
+	userName := nsTmplSet.GetName()
+	labels := map[string]string{
+		"owner": userName,
```

Comment:
could we please use constants for such strings? applies also for the following ones

Reply:
Sometimes direct const make things more readable.  I have seen such practice followed in other places in our code too.